Modern UI - Ad Screen ControlBar adjustments

123mpozzi opened the pull request:
## Description
<!-- Add a short description about the changes -->

This PR provides some UX adjustments to the control bar shown in ad layouts.

## Changes
- Slide down **background shadow** on controls hide
  > Gives more space to the ad itself
- Keep `AdSkipButton` in a fixed position
  > Since sliding it down on controls hide has worse UX: what if the
user is pointing with the mouse or finger at the button, and the
button goes down at that same moment?
Keeping it fixed position will ensure the user does not have to
adjust their pointer position again
- Refactor `AdControlBarBottom` and part of `ControlBar` logic into its own `AdControlBar` to better separate the concerns

## Checklist (for PR submitter and reviewers)
<!-- A PR with CHANGELOG entry will be released automatically -->
<!-- This is required and should be added in every case -->
- [x] `CHANGELOG` entry - `no need since base is modern/ui, a changelog will be added at the end release`

123mpozzi committed:
Create `AdControlBar` to hold ad logic instead of `AdControlBarBottom`

123mpozzi committed:
Fix show/hide not applying slide animations as expected

123mpozzi committed:
Fix bottom part of `AdControlBar` being hidden by default

123mpozzi committed:
Slide down the whole control bar altogether instead of single elements

123mpozzi committed:
Only slide down AdControlBar with bottom component. Add to TV ad layout

123mpozzi committed:
Make `AdSkipButton` fixed position

Since sliding it down on controls hide has worse UX: what if the
user is pointing with the mouse or finger at the button, and the
button goes down at that same moment?

Keeping it fixed position will ensure the user does not have to
adjust their pointer position again.

123mpozzi committed:
Improve docs

123mpozzi committed:
Move default AdControlBar css classes to its constructor

123mpozzi committed:
Remove comment

stonko1994 commented on src/ts/components/ads/AdControlBar.ts:
```diff
@@ -0,0 +1,96 @@
+export interface AdControlBarConfig extends ControlBarConfig {
+  topComponents?: Container<ContainerConfig>;
+  bottomComponents?: Container<ContainerConfig>;
+}
```
I'm not sure if this is a good API design if we add this top and bottom components to the Config. The reason is that the `ControlBarConfig` still accepts the `components` attribute as well. This would be quite confusing having all 3. `components`, `topComponents` and `bottomComponents`. Is there a ways to get the same functionality working without the need for this two new properites?

stonko1994 replied:
Idea: https://github.com/bitmovin/bitmovin-player-ui/pull/736#discussion_r2282574875

123mpozzi replied:
Switched to dynamically computer top/bottom components in b432ff0c0f5bfe0562f05cae976c1642c1407d62 .
Now `AdControlBar` is used via `components` like a normal element

stonko1994 commented on src/scss/components/ads/_ad-control-bar.scss:
```diff
@@ -2,8 +2,33 @@
+  background: linear-gradient(to bottom, $color-transparent, $color-background-bars);
+  box-sizing: border-box;
+  line-height: 1em;
+  padding: 1em 1em .5em;
```
I changd the bars to display flex in [this](https://github.com/bitmovin/bitmovin-player-ui/pull/733/files) PR. Can you do the same for this new bar?

123mpozzi replied:
Sure, the design is based on the normal ControlBar so no issues :)
Address in 3fe4c07a054802a644895ea60fe241e2f850607a

stonko1994 commented on src/scss/components/ads/_ad-status-overlay.scss:
Is the `AdStatusOverlay` still used?

123mpozzi replied:
Yes, the control bar does not contain the skip button itself. That is still inside `AdStatusOverlay`

stonko1994 commented on src/ts/components/ads/AdControlBar.ts:
```diff
@@ -0,0 +1,96 @@
+ * @category Components
+ */
+export class AdControlBar extends ControlBar {
+  private static readonly CLASS_SLID_DOWN = 'slid-down';
```
Intended to use past tens for this class? `slid` vs `slide`?

123mpozzi replied:
Yes, the past tense is intentional. The idea is to indicate that the transition has already been applied.
So `slid` means the `slide-down` effect has been applied and is currently in effect.

stonko1994 commented on src/ts/components/ads/AdControlBar.ts:
```diff
@@ -0,0 +1,96 @@
+  }
+
+  hide(): void {
+    if (this.bottomContainer) {
```
What if we recursively check all components in `config.components` and look for the seekbar and hide all other containers? 🤔 Too much magic?

123mpozzi committed:
Switch to flexbox to handle spacing

As we did for ControlBar in
https://github.com/bitmovin/bitmovin-player-ui/pull/733/files

123mpozzi committed:
Remove top/bottom components, and dynamically infer them instead

123mpozzi committed:
Switch few for loops to functional alternatives

123mpozzi committed:
Revert switches of ticks

123mpozzi committed:
Add `AdControlBar`

stonko1994 commented on src/ts/UIFactory.ts:
```diff
@@ -356,11 +356,16 @@ function adsUILayout() {
+          new Spacer(),
+          new FullscreenToggleButton(),
+        ],
+      }),
```
Please keep the previous approach, where it's possible to set CSS classes from here. This allows more flexibility in case someone wants to build his own UI layout. **Applies for all occasions**

```suggestion
      new Container({
        components: [
          new PlaybackToggleButton(),
          new VolumeToggleButton(),
          new Spacer(),
          new FullscreenToggleButton(),
        ],
        cssClasses: ['controlbar-top ad-controlbar-top'],
      }),
```

123mpozzi replied:
Address in d2e834f0831c578144c93616a53a8527a5bdb30e

stonko1994 commented on src/ts/components/ads/AdControlBar.ts:
```diff
@@ -0,0 +1,115 @@
+  // Uses the inherited 'components' property from ControlBarConfig
+  // Components containing SeekBar will be treated as "top" components
+  // Components without SeekBar will be treated as "bottom" components
+}
```
I think we don't need this at all if there is no additional config option.

123mpozzi replied:
Address in d2e834f0831c578144c93616a53a8527a5bdb30e

stonko1994 reviewed (commented):
I think we need to rethink the whole slide down solution. Customization of our UI is a big aspect and it should be easo to support the same feature also e.g. when the seekbar is in the bottom container of the `AdControlBar` or if there is only one `Container`.

Let's talk offline.

123mpozzi committed:
Allow AdControlBar custom classes. Collapse height for hide animation

123mpozzi commented:
> I think we need to rethink the whole slide down solution. Customization of our UI is a big aspect and it should be easo to support the same feature also e.g. when the seekbar is in the bottom container of the `AdControlBar` or if there is only one `Container`.
> 
> Let's talk offline.

I address this in d2e834f0831c578144c93616a53a8527a5bdb30e

123mpozzi committed:
Ignore sass lint rule for a double animation (height, opacity)

123mpozzi committed:
Rename and re-order local mixin

stonko1994 commented on src/scss/components/ads/_ad-control-bar.scss:
```diff
@@ -0,0 +1,67 @@
+// Keep this mixin here for now since it is specific to the ad control bar
+@mixin hidden-animated-collapse-and-fade-out($duration: $animation-duration, $original-height: 2em) {
+  // sass-lint:disable no-transition-all
+  transition: all $duration ease-out;
```
Is the `transition: all` necessary? can we specify opacity and height explicitly?

123mpozzi replied:
Yes, we can. Address in ae178d099f3fdb1cf217f504d6f1b68f326ced4c

stonko1994 commented on src/ts/components/ads/AdControlBar.ts:
```diff
@@ -0,0 +1,120 @@
+    super({
+      ...config,
+      cssClasses: ['ad-controlbar'],
+    });
```
Can we keep our current style of calling super in components: (and see comment below)


```suggestion
    super(config);
```

123mpozzi replied:
Address in 75d263672ddb3d9313b927396e09d3156d252636

stonko1994 commented on src/ts/components/ads/AdControlBar.ts:
```diff
@@ -0,0 +1,120 @@
+    });
+
+    this.config = this.mergeConfig(config, {
+      cssClass: 'ui-controlbar',
```
Let's use the cssClasses approach to add additional classes to this component:


```suggestion
      cssClasses: ['ad-controlbar'],
```

123mpozzi replied:
Address in 75d263672ddb3d9313b927396e09d3156d252636

stonko1994 commented on src/ts/components/ads/AdControlBar.ts:
```diff
@@ -0,0 +1,120 @@
+    });
+  }
+
+  private findContainerWithSeekBar(container: Container<ContainerConfig>): boolean {
```
The naming and the return value doesn't match IMO. The naming of `findContainer` would indicate that the return value is a `Container`. Which we could do as well which could also simplify its implementaiton.

A return type of boolean rather indicates that the name should be `containsSeekBar` or similar.

123mpozzi replied:
Make sense, address in 45951a81c21ae67071fd4da170376364428d30bd

stonko1994 commented on src/ts/UIFactory.ts:
```diff
@@ -345,9 +345,9 @@ function uiLayout(config: UIConfig) {
+  let controlBar = new AdControlBar({
     components: [
-      new Container({
+      new Container(<ContainerConfig>{
```
I think we can remove the type cast here


```suggestion
      new Container({
```

stonko1994 commented on src/ts/UIFactory.ts:
```diff
@@ -356,11 +356,19 @@ function adsUILayout() {
+        cssClasses: ['ad-controlbar-top'],
+
+      }),
+      new Container(<ContainerConfig>{
```
I think we can remove the type cast here


```suggestion
      new Container({
```

123mpozzi replied:
Address in 895ed81d36d15421f116395c4ba261f0cec44621

stonko1994 commented on src/ts/UIFactory.ts:
```diff
@@ -513,9 +521,9 @@ function smallScreenUILayout() {
+  let controlBar = new AdControlBar({
     components: [
-      new Container({
+      new Container(<ContainerConfig>{
```
```suggestion
      new Container({
```

stonko1994 commented on src/ts/UIFactory.ts:
```diff
@@ -524,11 +532,18 @@ function smallScreenAdsUILayout() {
-        cssClasses: ['controlbar-top ad-controlbar-top'],
+        cssClasses: ['ad-controlbar-top'],
+      }),
+      new Container(<ContainerConfig>{
```
```suggestion
      new Container({
```

stonko1994 commented on src/ts/UIFactory.ts:
```diff
@@ -697,9 +712,9 @@ function tvAdsUILayout() {
+      new AdControlBar({
         components: [
-          new Container({
+          new Container(<ContainerConfig>{
```
```suggestion
          new Container({
```

stonko1994 reviewed (commented):
Getting there 💪 Do you know where this padding (purple) is coming from? Makes the seekbar look out of place slightly. <img width="446" height="55" alt="Image" src="https://github.com/user-attachments/assets/7d72df9e-0afe-47fe-9e79-3e636d0872b7" />

123mpozzi committed:
Merge branch 'feature/modern-ui-base' into feature/modern-ui-ad-screen-slide-shadow-background-as-well

123mpozzi committed:
Specify animations during transitions instead of generic `all`

123mpozzi committed:
Move classes setup to `mergeConfig`

123mpozzi committed:
Rename contains seekBar helper

123mpozzi commented:
> Getting there 💪 Do you know where this padding (purple) is coming from? Makes the seekbar look out of place slightly. <img alt="Image" width="446" height="55" src="https://private-user-images.githubusercontent.com/6216959/480375977-7d72df9e-0afe-47fe-9e79-3e636d0872b7.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTU3NjE0MTEsIm5iZiI6MTc1NTc2MTExMSwicGF0aCI6Ii82MjE2OTU5LzQ4MDM3NTk3Ny03ZDcyZGY5ZS0wYWZlLTQ3ZmUtOWU3OS0zZTYzNmQwODcyYjcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDgyMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA4MjFUMDcyNTExWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Mjg5NDkzODEyZDkzNzllZTc0MjU2Yzk0YmRiYTM3MDQ4ZjA4ODkyZWI0MzA4MGQ2MDBlYTczM2QyN2RhMGJmYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.qLPM027X350Ft_MMdIrzDsPBDKkQpWjmfrTIu8MaS14">

This is coming from `row-gap: 0.5em` and was introduced in 3fe4c07a054802a644895ea60fe241e2f850607a to address [this comment](https://github.com/bitmovin/bitmovin-player-ui/pull/736#discussion_r2282569437).

The goal of that commit was to style the ad controlbar the same way controlbar is styled in `feature/modern-ui-base` since it was changed in #733.

The style is the exact same one used for the normal controlbar.

123mpozzi committed:
Remove explicit type casting where unnecessary

stonko1994 commented on src/ts/components/ads/AdControlBar.ts:
```diff
@@ -0,0 +1,117 @@
+    super(config);
+
+    this.config = this.mergeConfig(config, {
+      cssClasses: ['ui-controlbar', 'ad-controlbar'],
```
The `ui-controlbar` class is already added by the super class so you don't have to add it again. This way it results in adding the same class twice:

<img width="618" height="22" alt="Image" src="https://github.com/user-attachments/assets/88f839ad-1980-4f38-9e82-99e380a38973" />


```suggestion
      cssClasses: ['ad-controlbar'],
```

stonko1994 commented:
> This is coming from `row-gap: 0.5em` and was introduced in [3fe4c07](https://github.com/bitmovin/bitmovin-player-ui/commit/3fe4c07a054802a644895ea60fe241e2f850607a) to address [this comment](https://github.com/bitmovin/bitmovin-player-ui/pull/736#discussion_r2282569437).
> 
> The goal of that commit was to style the ad controlbar the same way controlbar is styled in `feature/modern-ui-base` since it was changed in #733.
> 
> The style is the exact same one used for the normal controlbar.

Ah I see, 💩. Do you have an idea how we could fix this problem?

123mpozzi committed:
Remove duplicated css class

Co-authored-by: David Steinacher <david.steinacher@bitmovin.com>

stonko1994 commented on src/scss/components/ads/_ad-control-bar.scss:
```diff
@@ -0,0 +1,66 @@
+  }
+}
+
+.#{$prefix}-ui-controlbar {
```
This should be `ad-controlbar` no? The styles in here are specific for the ad control bar and shouldn't affect the regular `ui-controlbar` as far as I know.


```suggestion
.#{$prefix}-ad-controlbar {
```

123mpozzi replied:
Yes, let's do that

123mpozzi committed:
Do not duplicate styles for generic control bar

Co-authored-by: David Steinacher <david.steinacher@bitmovin.com>

123mpozzi commented:
> > This is coming from `row-gap: 0.5em` and was introduced in [3fe4c07](https://github.com/bitmovin/bitmovin-player-ui/commit/3fe4c07a054802a644895ea60fe241e2f850607a) to address [this comment](https://github.com/bitmovin/bitmovin-player-ui/pull/736#discussion_r2282569437).
> > The goal of that commit was to style the ad controlbar the same way controlbar is styled in `feature/modern-ui-base` since it was changed in #733.
> > The style is the exact same one used for the normal controlbar.
> 
> Ah I see, 💩. Do you have an idea how we could fix this problem?

I have a few, let's continue with this on another PR tho
